sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -66,6 +66,23 @@ private[execution] sealed trait HashedRelation extends KnownSizeEstimation {
     throw new UnsupportedOperationException
   }
 
+  /**
+   * Returns key index and matched rows.
+   *
+   * Returns null if there is no matched rows.
+   */
+  def getWithKeyIndex(key: InternalRow): (Int, Iterator[InternalRow])
+
+  /**
+   * Returns an iterator for keys index and rows of InternalRow type.
+   */
+  def valuesWithKeyIndex(): Iterator[(Int, InternalRow)]
```

Comment:
I'm a bit worried about creating `Tuple2` objects frequently. Maybe we can add a
```
class ValueRowWithKeyIndex {
  var keyIndex: Int
  var value: InternalRow
}
```

So that we can reuse the instance.